batch/batch/driver/k8s_cache.py

```diff
@@ -0,0 +1,55 @@
+import time
+import asyncio
+import logging
+
+log = logging.getLogger('k8s_cache')
+
+
+class K8sCache:
+    def __init__(self, client, refresh_time):
+        self.client = client
+        self.refresh_time = refresh_time
+        self.secrets = {}
+        self.service_accounts = {}
+
+    async def async_init(self):
+        asyncio.ensure_future(self.cleanup_loop())
+
+    async def read_secret(self, name, namespace, timeout):
+        secret, time_updated = self.secrets.get((name, namespace), (None, None))
+        if time_updated and time.time() < time_updated + self.refresh_time:
+            return secret
+
+        secret = await self.client.read_namespaced_secret(
+            name, namespace, _request_timeout=timeout)
+        self.secrets[(name, namespace)] = (secret, time.time())
+        return secret
+
+    async def read_service_account(self, name, namespace, timeout):
+        sa, time_updated = self.service_accounts.get((name, namespace), (None, None))
+        if time_updated and time.time() < time_updated + self.refresh_time:
+            return sa
+
+        sa = await self.client.read_namespaced_service_account(
+            name, namespace, _request_timeout=timeout)
+        self.service_accounts[(name, namespace)] = (sa, time.time())
+        return sa
+
+    async def cleanup_loop(self):
```

Comment:
It's probably fine, but this doesn't bound the size of the cache.  I'd probably just fix the size of the cache and remove the oldest element(s) (use a sorted dict) when adding a new element if it is full.

Reply:
I couldn't figure out how to use sorted dict, so I implemented it with a regular dict / sorted set.